File: 0047-permutations-ii/0047-permutations-ii.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permuteUnique = function (nums) {
    const dfs = function (choices, path) {
        if (path.length === nums.length) {
            result.push([...path]);
            return;
        }

        for (let i = 0; i < choices.length; i++) {
            if (i > 0 && choices[i - 1] === choices[i]) continue;
            const newChoices = [...choices.slice(0, i), ...choices.slice(i + 1)];
            path.push(choices[i]);
            dfs(newChoices, path);
            path.pop();
        }
    }

    nums.sort((a, b) => a - b);
    const result = [];
    dfs(nums, []);

    return result;
};

File: 0047-permutations-ii/0047-permutations-ii.py
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        def dfs(choices, path):
            if len(path) == len(nums):
                result.append(path[:])
                return

            for i in range(len(choices)):
                if i > 0 and choices[i - 1] == choices[i]:
                    continue

                newChoices = choices[0:i] + choices[i + 1 :]
                path.append(choices[i])
                dfs(newChoices, path)
                path.pop()

        nums.sort()
        result = []
        dfs(nums, [])

        return result
